### PR TITLE
[merge conflict] add search_doc()

### PR DESCRIPTION
**Pull request by DrXyzzy**
This code is for the `search_doc` part of #724 .

Tests in sample worksheet in project with feature installed here:
https://cloud.sagemath.com/projects/ccd4d4a4-29a8-4c39-85c2-a630cb1e9b6c/files/SAGEWS/search_doc_724_test.sagews

### CODE REVIEW

**Comment by DrXyzzy** [comment]
Added code for `search_src` to this PR.

**Comment by GregoryBard** [comment]
I tried to test this, but it didn't work for me. At the first "search_doc"
I got the following error message:

Error in lines 1-1Traceback (most recent call last):
  File "/projects/ccd4d4a4-29a8-4c39-85c2-a630cb1e9b6c/.local/lib/python2.7/site-packages/smc_sagews/sage_server.py",
line 947, in execute
    exec compile(block+'\n', '', 'single') in namespace, locals
  File "", line 1, in <module>
  File "/projects/sage/sage-6.10/local/lib/python2.7/site-packages/sage/misc/sagedoc.py",
line 1139, in search_doc
    **kwds)
  File "/projects/sage/sage-6.10/local/lib/python2.7/site-packages/sage/misc/sagedoc.py",
line 925, in _search_src_or_doc
    pager.pager()(results)
  File "/projects/ccd4d4a4-29a8-4c39-85c2-a630cb1e9b6c/.local/lib/python2.7/site-packages/IPython/core/page.py",
line 273, in page
    return pager_page(data, start, screen_lines, pager_cmd)
  File "/projects/ccd4d4a4-29a8-4c39-85c2-a630cb1e9b6c/.local/lib/python2.7/site-packages/IPython/core/page.py",
line 233, in pager_page
    pager_encoding = pager.encoding or sys.stdout.encoding
AttributeError: 'BufferedOutputStream' object has no attribute 'encoding'


---Greg

**Comment by DrXyzzy** [comment]
I had changed the smc branch in that project to work on other issues. Should be good to go now. You can confirm current branch if you open greg.term in the project and do cd ~/smc; git branch. Should match the branch name at the top of the PR.

**Comment by haraldschilly** [comment]
ok, the test page works as advertised, I'm going to read the diff now

File: src/smc_sagews/smc_sagews/sage_salvus.py [hunk 1]
```diff
@@ -3612,6 +3612,40 @@ def help(*args, **kwds):
...
+    INPUT:
+
+    - ``str`` -- string expression to search for; will be quoted
+
+    OUTPUT:
+
+    Interact selector of matching filenames. Choosing one causes its
+    contents to be shown in salvus.code() output.
+    """
+    cmd = "cd /projects/sage/sage/src/sage;git grep -il '" + str + "'"
```
Comment (haraldschilly) [inline src/smc_sagews/smc_sagews/sage_salvus.py]: in the new image, sage will be somewhere else. use the `SAGE_ROOT` env variable in `os.environ` ... and default it to that path there: `os.environ.get('SAGE_ROOT', '/.../...')`

Comment (haraldschilly) [inline src/smc_sagews/smc_sagews/sage_salvus.py]: ok ,and I just realize, there is no `SAGE_ROOT`. uhm...

Comment (haraldschilly) [inline src/smc_sagews/smc_sagews/sage_salvus.py]: ok, going the ridiculous way: `cd $(dirname $(readlink -f $(which sage))); ...`

Comment (haraldschilly) [inline src/smc_sagews/smc_sagews/sage_salvus.py]: and another idea for grep: prepend it with `timeout 5 grep -il ...` such that it times out after 5 secs and you still get the output. Maybe 10 secs? IDK.

**Comment by haraldschilly** [comment]
ok, made some comments to make it independent of the hardcoded path and maybe add the timeout. apart from that +1

**Comment by DrXyzzy** [comment]
Thanks. Changed issue title to **in progress**.

**Comment by DrXyzzy** [comment]
Logic for path to sage executable altered and timeout added per @haraldschilly suggestion above. Also added max_chars optional parameter to search_src().

**Comment by haraldschilly** [comment]
hmm, now it complains about merge conflicts. that's odd, where there changes in those files?

**Pull request merged**